samza-core/src/main/java/org/apache/samza/elasticity/ElasticityUtils.java

```diff
@@ -0,0 +1,486 @@
+   *    we have getSystemStreamPartition(int elasticityFactor) which does keyBucket = (Math.abs(envelopeKeyorOffset.hashCode()) % 31) % elasticityFactor;
+   *    Thus,
+   *       SSP = SSP_withKeyBucket0 + SSP_withKeyBucket1.
+   *       SSP_withKeyBucket0 = SSP_withKeyBucket00 + SSP_withKeyBucket01.
+   *    If the checkpoint map has
+   *      Partition 0: (SSP : 1), Partition 0_0_2: (SSP0 : 2), Partition 0_1_2: (SSP1 : 3), Partition 0_0_4: (SSP0 : 4), Partition 0_2_4: (SSP1 : 6)
+   *      looking at these map and knowing that offsets are monotonically increasing, it is clear that last deploy was with elasticity factor = 4
+   *      to get checkpoint for Partition 0_0_2, we need to consider last deploy's offsets.
+   *      picking 6 (offset for Partition 0_2_4) means that 0_0_2 will start proc from 6 but offset 5 was never processed.
+   *      hence we need to take min of offsets within an elasticity factor.
+   *
+   * Given checkpoints for all the tasks in the checkpoint stream,
+   * computing the last proc offset for an ssp checkpoint for a task,
+   * the following needs to be met.
+   *    1. Ancestors: we need to take largest offset among ancestors for an ssp
+   *    2. Descendants:
+   *         a. group descendants by their elasticityFactor.
+   *         b. among descendants of the same elasticityFactor, take the smallest offset for an ssp
+   *         c. once step b is done, we have (elasticityFactor : smallest-offset-for-ssp) set, pick the largest in this set
+   *    3. Pick the larger among the offsets received from step 1 (for ancestors) and step 2 (for descendants)
+   *
+   * @param taskName
+   * @param taskSSPSet
+   * @param checkpointMap
+   * @param systemAdmins
+   * @return
+   */
+  public static Map<SystemStreamPartition, String> computeLastProcessedOffsetsFromCheckpointMap(
+      TaskName taskName,
+      Set<SystemStreamPartition> taskSSPSet,
+      Map<TaskName, Checkpoint> checkpointMap,
+      SystemAdmins systemAdmins) {
+    Pair<Set<Checkpoint>, Map<Integer, Set<Checkpoint>>> acnestorsAndDescendantsFound =
+        getAncestorAndDescendantCheckpoints(taskName, checkpointMap);
+    Set<Checkpoint> ancestorCheckpoints = acnestorsAndDescendantsFound.getLeft();
+    Map<Integer, Set<Checkpoint>> descendantCheckpoints = acnestorsAndDescendantsFound.getRight();
+
+    Map<SystemStreamPartition, String> taskSSPOffsets = new HashMap<>();
+
+    taskSSPSet.forEach(ssp_withKeyBucket -> {
+      log.info("for taskName {} and ssp of the task {}, finding its last proc offset", taskName, ssp_withKeyBucket);
+
+      SystemStreamPartition ssp = new SystemStreamPartition(ssp_withKeyBucket.getSystemStream(),
+          ssp_withKeyBucket.getPartition());
+
+      SystemAdmin systemAdmin = systemAdmins.getSystemAdmin(ssp.getSystem());
+
+      String currentLastOffsetForSSP = null;
+
+      String ancestorLastOffsetForSSP = getMaxOffsetForSSPInCheckpointSet(ancestorCheckpoints, ssp, systemAdmin);
```

Comment:
qq: what's the use case for combining several ancestors? If we increase the elasticity factor, looks like we only need to assign the offset of the current ancestor to its descendants. And if we decrease the elasticity factor, we only need to combine the offsets of several descendants using the min function.